src/distributions/uniform.rs

```diff
@@ -187,68 +188,51 @@ impl Distribution<bool> for Uniform {
-    /// an affine transition
-    /// http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/ARTICLES/dSFMT.pdf
-    /// http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/SFMT/dSFMT-slide-e.pdf
-    fn sample<R: Rng+?Sized>(&self, rng: &mut R) -> f32 {
-        const UPPER_MASK: u32 = 0x3F800000;
-        const LOWER_MASK: u32 = 0x7FFFFF;
-        let tmp = UPPER_MASK | (rng.next_u32() & LOWER_MASK);
-        let result: f32 = unsafe { mem::transmute(tmp) };
-        result - 1.0
-    }
-}
-
-impl Distribution<f64> for Uniform01 {
-    fn sample<R: Rng+?Sized>(&self, rng: &mut R) -> f64 {
-        const UPPER_MASK: u64 = 0x3FF0000000000000;
-        const LOWER_MASK: u64 = 0xFFFFFFFFFFFFF;
-        let tmp = UPPER_MASK | (rng.next_u64() & LOWER_MASK);
-        let result: f64 = unsafe { mem::transmute(tmp) };
-        result - 1.0
-    }
-}
-
 macro_rules! float_impls {
-    ($scale_name:ident, $ty:ty, $mantissa_bits:expr) => {
-        const $scale_name: $ty = (1u64 << $mantissa_bits) as $ty;
-        
-        impl Distribution<$ty> for Open01 {
-            #[inline]
+    ($ty:ty, $next_u:path) => {
+        impl Distribution<$ty> for Uniform01 {
             fn sample<R: Rng+?Sized>(&self, rng: &mut R) -> $ty {
-                // add a small amount (specifically 2 bits below
-                // the precision of f64/f32 at 1.0), so that small
-                // numbers are larger than 0, but large numbers
-                // aren't pushed to/above 1.
-                let x: $ty = Rand::rand(rng, Uniform01);
-                x + 0.25 / $scale_name
+                let x = $next_u(rng);
+                x.closed_open01()
             }
         }
+
         impl Distribution<$ty> for Closed01 {
-            #[inline]
             fn sample<R: Rng+?Sized>(&self, rng: &mut R) -> $ty {
-                // rescale so that 1.0 - epsilon becomes 1.0
-                // precisely.
-                let x: $ty = Rand::rand(rng, Uniform01);
-                x * $scale_name / ($scale_name - 1.0)
+                // The problem with a closed range over [0,1] is that it needs
+                // 2^n+1 samples to generate a uniform distribution. Which is
+                // difficult, as it means we either have to reject about half
+                // the generated random numbers, or just not include one number
+                // in the distribution. That is why, instead of a closed range,
+                // we actually sample from the half-open range (0,1].
+                //
+                // Floating-point numbers have more precision near zero. This
+                // means for a f32 that only 1 in 2^32 samples will give exactly
+                // 0.0. But 1 in 2^23 will give exactly 1.0 due to rounding.
+                // Because the chance to sample 0.0 is so low, this half-open
+                // range is a very good appoximation of a closed range.
+                let x = $next_u(rng);
+                x.open_closed01()
```

Comment:
You give a good rationale for implementing `[0, 1]` sampling via `(0, 1]` instead. I wonder what happens with your extra precision sampling under common transformations. E.g. if a user samples `x` via this function, then what's the actual range of `x + 1.0` — probably exactly `[1, 2]` since the loss of precision away from 0 forces rounding? Would `x - 1.0` have range `[-1, 0]`?

Reply:
Yes, scaling should mostly preserve the precision, but addition forces rounding. I hope the extra precision for `Open01` does not break anyone's code. But the most important reason to use `Open01` is to not get 0.0, for log, division, etc.. I don't imagine much problems in those cases.

`x - 1.0` should have `[-1, 0]` as range. I didn't think about that! But it hurts a bit now that I am focused on precision ;-)

Reply:
I'm wondering now whether we should permanently document `Closed01` as having range `(0, 1]`, or have another distribution like `OpenClosed01` as well, or instead of `Closed01`. Because not generating 0 may be a useful property.

Of course it may be surprising that `OpenClosed01 + x` will often have range `[x, x+1]` due to rounding, but there's not much we can do about that (not knowing in general how large x is).

Reply:
I would not hurry with adding more range distribution with a different closed/open story... We already have maybe to many to choose from. If someone doesn't want to generate 0.0, he can use `Open01`. And otherwise there is now the conversion function `open_closed01`. So I would not do add it right now, but who knows what the future brings?